### ISSUE TITLE
Spam Lists

### ISSUE THREAD

**Issue by bymayo** [michaelrog/Shortlist#23]
We've noticed since putting a site in to production that there are hundreds! (Growing in to thousands!) of wishlists with just solely 1 product in it, all with different users.

We've tested this a lot and can only put it down to some sort of spam bots hitting the add to wishlist links.

The lists are created when a user adds something to a list via the link option (Not a form) 

Is there anyway to combat this at all?

**Comment by michaelrog** [comment]
Just to be sure — Are you only making the link visible to logged-in users? (Otherwise, every search spider or spam bot in the world is gonna try to follow it.)

Also, can you tell whether those lists are owned by `member` or `guest` users? (The `ownerType` column in the db will say as much.)

**Comment by bymayo** [comment]
The lists are by `guest` users, as we want logged out users to create lists, then allow them to submit these lists as an actual member.

When you say only make the link visible to logged-in users, how is this possible if we want logged out users to create lists?

**Comment by michaelrog** [comment]
Unfortunately, if you need the link to be visible to guests, options for restricting its use are limited.

Adding a `nofollow` attribute to your links would probably help to cut down on _some_ of the automated traffic, though:

```<a href="..." rel="nofollow"> </a>```

You could also set a [_robots.txt_](https://developers.google.com/search/reference/robots_txt#url-matching-based-on-path-values) directive to [try to] disallow Shortlist links from being crawled:

```Disallow: /actions/shortlist```

**Comment by bymayo** [comment]
I'll try these and let you know if it works.

Is it not worth building in a function to detect if it's a spam link before it generates a list? I did this previously on a similar kind of plugin with good success using https://github.com/JayBizzle/Crawler-Detect

```
if(!$CrawlerDetect->isCrawler()) {
    // Not a bot!
}
```

**Comment by michaelrog** [comment]
That's not a bad suggestion. I'm not sure whether "spam protection" belongs as core functionality of Shortlist or in another coordinating plugin. Either way, we could definitely do more in that vein. Thanks for the link!

**Comment by bymayo** [comment]
Yeah! I think because it's an open link to a controller it needs a bit of protection. Look forward to seeing how you do this!

**Comment by bymayo** [comment]
@michaelrog  I hope you don't mind, but I forked the repo to add this functionality in temporarily until it's fixed. If anybody else needs this fix the commit is:

https://github.com/bymayo/Shortlist/commit/a403761dec493b124e963fd3032941d39437b72f

And obviously can be downloaded from my forked version. But this means it will be out of date with @michaelrog master. 

Once the fix is in, i'll remove this repo 👍

**Comment by michaelrog** [comment]
@bymayo Sure, it's fine to fork the repo in order to submit a PR. However, _Shortlist_'s [license](https://github.com/TopShelfCraft/Shortlist/blob/1.x/LICENSE.md) doesn't presently allow redistributing the code, so I'd appreciate if you would kindly _not_ encourage anyone to download code from you. (While unlikely, people having out-of-sync versions of the plugin could lead to support frustrations for me.) 😅 

Meanwhile, please do create a feature branch for your changes, PR to the `1.x.dev` branch, and reference this Issue. I'm grateful to you for taking the initiative to look into this, and I'll be happy to review your changes and see about merging them into the master source!

**Comment by bymayo** [comment]
Whoops! Sorry about that!

What i'll do then is create a feature branch and push to yours and remove my forked version 💯

We just had our client on our back you see! 😂
